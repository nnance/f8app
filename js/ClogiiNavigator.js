import React from 'react';
import {
  Linking,
} from 'react-native';
import { parse } from 'query-string';
import { createCommentContainer } from 'graphql-comment/src/react-native';
import gql from 'graphql-tag';

import FixBugScrollViewNavigator from './common/FixBugScrollViewNavigator';
import { withTracking } from './common/navigateTracking';

import ClogiiTabView from './tabs/ClogiiTabView';
import Reader from './reader';
import Book from './book';
const Comment = createCommentContainer({
  userOnCommentFragment: gql`
    fragment UserOnComment on Comment {
      author {
        id
        name
        profilePicture
      }
    }
  `,
  getAuthorOnComment: comment => comment.author ? comment.author : {},
  optimisticUserResponse: ({ ownProps }) => ({
    author: {
      __typename: 'User',
      id: 'unknow',
      name: 'unknow',
      profilePicture: null,
    },
  }),
});

class ClogiiNavigator extends React.Component {
  constructor(...args) {
    super(...args);
    this.onOpenURL = this.onOpenURL.bind(this);
    this.redirectTo = this.redirectTo.bind(this);
    this.parseURL = this.parseURL.bind(this);
    this.goToPlayer = this.goToPlayer.bind(this);
    this.goToComment = this.goToComment.bind(this);
    this.goToBook = this.goToBook.bind(this);
    this.goBack = this.goBack.bind(this);
    this.renderScene = this.renderScene.bind(this);
  }

  componentDidMount() {
    Linking.getInitialURL().then((url) => {
      if (url) {
        this.redirectTo(url);
      }
    }).catch((error) => {
      console.error('initial url error: ', error);
    });
    Linking.addEventListener('url', this.onOpenURL);
    // this.goToPlayer("312523187685666925511432");
  }

  componentWillUnmount() {
    Linking.removeEventListener('url', this.onOpenURL);
  }

  onOpenURL(event) {
    this.redirectTo(event.url);
  }

  redirectTo(url) {
    const { method, query } = this.parseURL(url);
    if (method === 'player' && query.id) {
      this.goToPlayer(query.id);
    }
    if (method === 'book' && query.id) {
      this.goToBook(query.id);
    }
  }

  parseURL(fUrl) {
    const url = fUrl;
    try {
      const liftedHost = url.split(':').length > 1 ? url.split(':')[1] : url.split(':')[0];
      const splited = liftedHost.split('?');
      let method = splited[0];
      const queryS = splited[1];
      const query = parse(queryS);
      if (method.indexOf('//') === 0) {
        method = method.slice(2);
      }
      return { method, query };
    } catch (error) {
      console.error('parse url error: ', error);
    }
    return {};
  }

  goToPlayer(id) {
    this.navigator.push({ page: 'player', id });
    this.props.navigate(`player?id=${id}`);
  }

  goToBook(id) {
    this.navigator.push({ page: 'book', id });
    this.props.navigate(`book?id=${id}`);
  }

  goToComment(id) {
    this.navigator.push({ page: 'comment', id });
    this.props.navigate(`comment?id=${id}`);
  }

  goBack() {
    this.navigator.pop();
    this.props.navigateBack();
  }

  renderScene(route) {
    if (route.page === 'main-tab') {
      return (<ClogiiTabView
        goToPlayer={this.goToPlayer}
        goToBook={this.goToBook}
        redirectTo={this.redirectTo}
      />);
    }
    if (route.page === 'player') {
      return (
        <Reader
          onBackPress={this.goBack}
          goToComment={this.goToComment}
          id={route.id}
        />
      );
    }
    if (route.page === 'book') {
      return (<Book
        onBackPress={this.goBack}
        goToPlayer={this.goToPlayer}
        goToComment={this.goToComment}
        id={route.id}
      />);
    }
    if (route.page === 'comment') {
      return (
        <Comment
          onBackPress={this.goBack}
          discussionRef={route.id}
        />
      );
    }
    return null;
  }

  render() {
    return (
      <FixBugScrollViewNavigator
        ref={
          (node) => {
            this.navigator = node;
          }
        }
        initialRoute={{ page: 'main-tab', id: 3 }}
        renderScene={this.renderScene}
      />
    );
  }
}

export default withTracking(ClogiiNavigator);
export {
  ClogiiNavigator as Component,
};
