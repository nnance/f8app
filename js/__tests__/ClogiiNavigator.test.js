/* eslint import/first: off */
jest.mock('react-native-fs', () => null);

import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { View } from 'react-native';

import { Component as ClogiiNavigator } from '../ClogiiNavigator';

function getWrapper(props = {}) {
  return shallow(<ClogiiNavigator {...props} navigate={jest.fn()} navigateBack={jest.fn()} />);
}

describe('ClogiiNavigator', () => {
  it('render', () => {
    const wrapper = getWrapper();
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('navigate to Book/Player', () => {
    const dump = getWrapper();
    const spy = jest.fn();
    dump.instance().navigator = {
      push: spy,
    };
    dump.instance().goToPlayer(5);
    expect(spy).toBeCalledWith({ id: 5, page: 'player' });
    dump.instance().goToBook(123);
    expect(spy).toBeCalledWith({ id: 123, page: 'book' });
  });

  it('navigate back', () => {
    const dump = getWrapper();
    const spy = jest.fn();
    dump.instance().navigator = {
      pop: spy,
    };
    dump.instance().goBack(5);
    expect(spy).toBeCalled();
  });

  it('onOpenURL will map event to redirectTo', () => {
    const dump = getWrapper();
    const spy = jest.fn();
    dump.instance().redirectTo = spy;
    dump.instance().onOpenURL({ url: 'x' });
    expect(spy).toBeCalledWith('x');
  });

  it('redirectTo will navigate to player on `clogii://player?id=123`', () => {
    const dump = getWrapper();
    const spy = jest.fn();
    dump.instance().navigator = {
      push: spy,
    };
    dump.instance().redirectTo('clogii://player?id=123');
    expect(spy).toBeCalledWith({ page: 'player', id: '123' });
  });

  it('parseURL split to method and query', () => {
    const dump = getWrapper();
    const result = dump.instance().parseURL('clogii://player?id=199');
    expect(result.method).toBe('player');
    expect(result.query).toEqual({ id: '199' });

    const result2 = dump.instance().parseURL('clogii://book?id=77&x=3');
    expect(result2.method).toBe('book');
    expect(result2.query).toEqual({ id: '77', x: '3' });
  });

  describe('renderScene', () => {
    it('main-tab', () => {
      const dump = getWrapper();
      const tree = shallow(<View>{dump.instance().renderScene({ page: 'main-tab' })}</View>);
      expect(toJSON(tree)).toMatchSnapshot();
    });

    it('player', () => {
      const dump = getWrapper();
      const tree = shallow(<View>{dump.instance().renderScene({ page: 'player', id: 1 })}</View>);
      expect(toJSON(tree)).toMatchSnapshot();
    });

    it('book', () => {
      const dump = getWrapper();
      const tree = shallow(<View>{dump.instance().renderScene({ page: 'book', id: 3 })}</View>);
      expect(toJSON(tree)).toMatchSnapshot();
    });
  });
});
