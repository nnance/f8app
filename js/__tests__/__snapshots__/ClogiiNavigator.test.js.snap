exports[`ClogiiNavigator render 1`] = `
<FixBugScrollViewNavigator
  initialRoute={
    Object {
      "id": 3,
      "page": "main-tab",
    }
  }
  renderScene={[Function]} />
`;

exports[`ClogiiNavigator renderScene book 1`] = `
<View>
  <Apollo(Book)
    goToComment={[Function]}
    goToPlayer={[Function]}
    id={3}
    onBackPress={[Function]} />
</View>
`;

exports[`ClogiiNavigator renderScene main-tab 1`] = `
<View>
  <Connect(Connect(ClogiiTabView))
    goToBook={[Function]}
    goToPlayer={[Function]}
    redirectTo={[Function]} />
</View>
`;

exports[`ClogiiNavigator renderScene player 1`] = `
<View>
  <Apollo(Apollo(Apollo(Player)))
    goToComment={[Function]}
    id={1}
    onBackPress={[Function]} />
</View>
`;
